libc-bottom-half/headers/public/wasix/context.h

```diff
@@ -0,0 +1,182 @@
+//
+// #### Wasmer implementation note
+// In the current Wasmer implementation, termination of the entrypoint always
+// resumes the main context and produces a trap:
+//
+// - Entrypoint terminated with a trap:
+//   The same trap is produced in the main context.
+// - Entrypoint returned normally:
+//   A custom user trap is produced in the main context.
+// - Entrypoint threw an uncaught exception:
+//   The exception is re-thrown in the main context.
+//
+// This behavior may change in future versions.
+//
+// #### Parameters
+// - `context_id` pointer to where the new context identifier will be stored
+// - `entrypoint` function invoked when the context is first resumed
+//
+// #### Return Value
+// - `0` on success
+// - `-1` on failure, with `errno` set
+//
+// #### Errors
+// - `EINVAL` entrypoint is not a valid function with the required signature
+// - `ENOTSUP` not in a context-switching environment
+// - `<memory-error>` Implementations may set `errno` to one of the
+// WASIX-defined
+//   memory error codes if `context_id` or memory reachable from the arguments
+//   is invalid or inaccessible.
+int wasix_context_create(wasix_context_id_t *context_id,
+                         void (*entrypoint)(void));
+
+// Destroy a suspended or terminated context.
+//
+// Destroys a context that is suspended or terminated. After a successful call,
+// the context enters the deleted state; its identifier becomes invalid and its
+// resources are released. Destroying an already deleted context is a no-op.
+//
+// The main context cannot be destroyed.
+//
+// Terminated contexts are effectively equivalent to deleted contexts, but
+// `wasix_context_destroy()` must still be called to release resources.
+//
+// #### Wasmer implementation note
+// Destroying a suspended context may produce an opaque user trap in the target
+// context, requiring the entrypoint to terminate. External functions that call
+// back into WASIX must propagate the trap unchanged; otherwise, behavior is
+// undefined.
+//
+// #### Parameters
+// - `context_id` identifier of the context to destroy
+//
+// #### Return Value
+// - `0` on success
+// - `-1` on failure, with `errno` set
+//
+// #### Errors
+// - `EINVAL` can not destroy the active context
+// - `EINVAL` can not destroy the main context
+// - `ENOTSUP` not in a context-switching environment
+int wasix_context_destroy(wasix_context_id_t context_id);
+
+// Suspend the active context and resume another.
+//
+// Suspends the active context and resumes execution in `target_context_id`.
+// The resumed context continues from where it was last suspended, or from its
+// entrypoint if it has never been resumed.
+//
+// If `target_context_id` refers to the active context, the call is a no-op.
+//
+// #### Wasmer implementation note
+// In the current Wasmer implementation, calling `wasix_context_switch()` in
+// the main context may produce a trap if another context terminates while the
```

Comment:
According to the doc comments, this is not wasmer-specific?

Reply:
wdym?